flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -281,15 +320,50 @@ protected void declineCheckpoint(long checkpointId) {
         public void run() {
             try {
                 mainOperator.run(lock, operatorChain);
-                if (!wasStoppedExternally && !isCanceled()) {
-                    synchronized (lock) {
-                        operatorChain.setIgnoreEndOfInput(false);
-                    }
-                }
+                completeProcessing();
                 completionFuture.complete(null);
             } catch (Throwable t) {
                 // Note, t can be also an InterruptedException
-                completionFuture.completeExceptionally(t);
+                if (isCanceled()
+                        && ExceptionUtils.findThrowable(t, InterruptedException.class)
+                                .isPresent()) {
+                    completionFuture.completeExceptionally(new CancelTaskException(t));
+                } else if (finishingReason == FinishingReason.STOP_WITH_SAVEPOINT_DRAIN
+                        && ExceptionUtils.findThrowable(t, InterruptedException.class)
+                                .isPresent()) {
+                    // if we are stopping the source thread for stop-with-savepoint
+                    // we may actually return from run with an InterruptedException which
+                    // should be ignored (e.g. Kinesis case see FLINK-23528)
+                    try {
+                        // clear the interrupted status for the thread
+                        Thread.interrupted();
+                        completeProcessing();
+                        completionFuture.complete(null);
+                    } catch (Throwable e) {
+                        completionFuture.completeExceptionally(e);
+                    }
```

Comment:
Do we need to support handle this case here? Can not we just ignore it and treat it as bug?

Reply:
I don't know... Why did we support it before?

Reply:
I presume it's a reminiscent of previous version of the code, where we were expecting failures/exceptions and we were interrupting the code. 

Can you maybe point to the code it was handling this case before?

Reply:
The next line in the current code:
```
 else if (finishingReason == FinishingReason.STOP_WITH_SAVEPOINT_NO_DRAIN) {
                    // swallow all exceptions if the source was stopped without drain
                    completionFuture.complete(null);
                }
```
or on master (https://github.com/apache/flink/blob/3b115544b04572831e162288097105c63ca5e048/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java#L183):
```
        sourceThread
                .getCompletionFuture()
                .whenComplete(
                        (Void ignore, Throwable sourceThreadThrowable) -> {
                            if (isCanceled()
                                    && ExceptionUtils.findThrowable(
                                                    sourceThreadThrowable,
                                                    InterruptedException.class)
                                            .isPresent()) {
                                mailboxProcessor.reportThrowable(
                                        new CancelTaskException(sourceThreadThrowable));
                            } else if (**!wasStoppedExternally** && sourceThreadThrowable != null) {
                                mailboxProcessor.reportThrowable(sourceThreadThrowable);
                            } else {
                                mailboxProcessor.suspend();
                            }
                        });
```

Reply:
Let's chat offline on Monday why has this code changed so much. I mean I don't understand what you mean that we did support it before. This branch:
```
finishingReason == FinishingReason.STOP_WITH_SAVEPOINT_DRAIN
                        && ExceptionUtils.findThrowable(t, InterruptedException.class)
                                .isPresent()
```
seems new to me.